### PR TITLE
Add spec for BatchNormTrainingOp

### LINKED ISSUE #346: Add spec for BatchNormTrainingOp
### Request description

To spec stablehlo.batch_norm_training

### PR DESCRIPTION
**Pull request by subhankarshah**
closes #346

### CODE REVIEW

File: docs/spec_draft.md [hunk 1]
```diff
@@ -343,6 +344,67 @@ logical operation.
...
+// %operand: [
+//            [[1.0, 2.0], [3.0, 4.0]],
+//            [[3.0, 4.0], [1.0, 2.0]]
+//           ]
+// %scale: [1.0, 1.0]
+// %offset: [1.0, 1.0]
+%results:3 = "mhlo.batch_norm_training"(%operand, %scale, %offset) {
+  epsilon = 0.0 : f32,
+  feature_index = 1 : i64
+} : (tensor<2x2x2xf32>, tensor<2xf32>, tensor<2xf32>, tensor<2xf32>, tensor<2xf32>) -> tensor<2x2x2xf32>
```
Comment (sdasgup3) [inline docs/spec_draft.md:397]: fix the arguments.

File: docs/spec_draft.md [hunk 2]
```diff
@@ -343,6 +344,67 @@ logical operation.
...
+| Name         | Type                                        |
+|--------------|---------------------------------------------|
+| `output`     | tensor of floating-point type               |
+| `batch_mean` | 1-dimensional tensor of floating-point type |
+| `batch_var`  | 1-dimensional tensor of floating-point type |
+
+### Constraints
+
+  * (C1) 0 $\le$ `feature_index` $\lt$ rank(`operand`).
+  * (C2) 0 $\lt$ rank(`operand`).
```
Comment (sdasgup3) [inline docs/spec_draft.md:378]: C2 redundant as it can be derived from C1.

File: stablehlo/dialect/StablehloOps.td [hunk 3]
```diff
@@ -1607,9 +1607,20 @@ def StableHLO_BatchNormTrainingOp : StableHLO_Op<"batch_norm_training",
...
+    See:
+    https://github.com/openxla/stablehlo/blob/main/docs/spec_draft.md#stablehlobatch_norm_training
 
-    See https://www.tensorflow.org/xla/operation_semantics#batchnormtraining
+    Example:
+    ```mlir
+    %results:3 = "mhlo.batch_norm_training"(%operand, %scale, %offset) {
+      epsilon = 0.0 : f32,
+      feature_index = 1 : i64
+    } : (tensor<2x2x2xf32>, tensor<2xf32>, tensor<2xf32>, tensor<2xf32>, tensor<2xf32>) -> tensor<2x2x2xf32>
```
Comment (sdasgup3) [inline stablehlo/dialect/StablehloOps.td:1622]: fix arguments.

File: docs/spec_draft.md [hunk 4]
```diff
@@ -343,6 +344,66 @@ logical operation.
...
 
+## stablehlo.batch_norm_training
+
+### Semantics
+
+Computes mean and variance across batch and spatial dimensions and normalizes
+the `operand` tensor, for each feature in the `feature_index` dimension and
+produces `output`, `batch_mean` and `batch_var` tensors. Refer
+[Batch Normalization Paper](https://arxiv.org/abs/1502.03167) for the detailed
+algorithm.
```
Comment (subhankarshah) [inline docs/spec_draft.md:355]: image

How about saying this Batch Norm Transform from the paper

Comment (subhankarshah) [reply docs/spec_draft.md:356]: @sdasgup3 reworded the semantics

**Review by sdasgup3 — APPROVED** [review]

**Comment by sdasgup3** [comment]
As per our conversation, this op will be referred to by the BatchNorm Inference highlighting the changes needed.

File: docs/spec_draft.md (hunk 4 above)
Comment (burmako) [reply docs/spec_draft.md:356]: As discussed offline, the current plan is to reformulate the Semantics section similarly to `batch_norm_inference`: https://github.com/openxla/stablehlo/pull/377#discussion_r1013499398. I'll wait for the reformulation before reviewing the section.

File: docs/spec_draft.md [hunk 5]
```diff
@@ -343,6 +344,67 @@ logical operation.
...
+
+| Name         | Type                                        |
+|--------------|---------------------------------------------|
+| `output`     | tensor of floating-point type               |
+| `batch_mean` | 1-dimensional tensor of floating-point type |
+| `batch_var`  | 1-dimensional tensor of floating-point type |
+
+### Constraints
+
+  * (C1) 0 $\le$ `feature_index` $\lt$ rank(`operand`).
```
Comment (burmako) [inline docs/spec_draft.md:378]: Similarly to https://github.com/openxla/stablehlo/pull/377#discussion_r1013442113, let's add another constraint here that says:

```
  * (C2) `operand`, `scale`, `offset`, `mean`, variance` and `result` have the 
    same element type.
```

File: docs/status.md [hunk 6]
```diff
@@ -42,7 +42,7 @@ one of the following tracking labels.
 | atan2                    | no            | yes*         | yes*           | yes             | no          |
 | batch_norm_grad          | no            | yes*         | yes*           | no              | no          |
 | batch_norm_inference     | no            | yes*         | yes*           | no              | no          |
-| batch_norm_training      | no            | yes*         | yes*           | no              | no          |
+| batch_norm_training      | yes           | yes          | yes            | no              | no          |
```
Comment (burmako) [inline docs/status.md:45]: Similarly to https://github.com/openxla/stablehlo/pull/377#discussion_r1013445041, let's change verification to `revisit` while we're figuring out #369.

File: docs/spec_draft.md [hunk 7]
```diff
@@ -343,6 +344,67 @@ logical operation.
...
+%results:3 = "stablehlo.batch_norm_training"(%operand, %scale, %offset) {
+  epsilon = 0.0 : f32,
+  feature_index = 2 : i64
+} : (tensor<2x2x2xf32>, tensor<2xf32>, tensor<2xf32>) -> (tensor<2x2x2xf32>, tensor<2xf32>, tensor<2xf32>)
+// %results#0: [
+//              [[0, 0], [2, 2]],
+//              [[2, 2], [0, 0]]
+//             ]
+// %results#1: [2, 3]
+// %results#2: [1, 1]
```
Comment (burmako) [inline docs/spec_draft.md:403]: Similarly to https://github.com/openxla/stablehlo/pull/377#discussion_r1013464149, since we're using floating-point notation above, let's use it for the example output as well.

Comment (burmako) [inline docs/spec_draft.md:403]: Let's use the same example across `batch_norm_inference` and `batch_norm_training` to emphasize how closely they are related. More concretely, let's take the example for `batch_norm_training` and copy/paste its operand, scale, offset, output, batch_var and batch_mean into `batch_norm_inference`.

**Review by burmako — CHANGES_REQUESTED** [review]

File: docs/spec_draft.md (hunk 7 above)
Comment (subhankarshah) [reply docs/spec_draft.md:430]: The outputs of the interpreter were without decimal point so I removed to be precise, will add it again.

File: docs/spec_draft.md [hunk 8]
```diff
@@ -348,6 +349,93 @@ logical operation.
...
+  # Compute `mean` and `variance` (as 1-dimensional tensors)
+  mean = compute_mean(operand, feature_index)
+  variance = compute_variance(operand, feature_index)
+
+  # Perform normalization with computed `mean` and `variance` instead of
+  # depending on `mean` and `variance` provided in the input like
+  # `batch_norm_inference`
+  return batch_norm_inference(operand, scale, offset, mean,
+                              variance, epsilon, feature_index)
+```
```
Comment (burmako) [inline docs/spec_draft.md:383]: Proposed refactoring of the code:

```Python
def compute_mean(operand, feature_index):
  (sum,) = reduce(
      inputs=[operand],
      init_values=[0],
      dimensions=[i for i in range(rank(operand)) if i != feature_index],
      body=lambda x, y: add(x, y))
  divisor = constant(num_elements(operand) / dim(operand, feature_index))
  divisor_bcast = broadcast_in_dim(mean, [], shape(sum))
  return divide(sum, divisor_bcast)

def compute_variance(operand, feature_index):
  mean = compute_mean(operand, feature_index)
  mean_bcast = broadcast_in_dim(mean, [feature_index], shape(operand))
  centered_operand = subtract(operand, mean_bcast)
  return compute_mean(mul(centered_operand, centered_operand), feature_index)

def batch_norm_training(operand, scale, offset, epsilon, feature_index):
  mean = compute_mean(operand, feature_index)
  variance = compute_variance(operand, feature_index)
  return batch_norm_inference(operand, scale, offset, mean, variance, epsilon,
                              feature_index)
```

Feedback:
  1. For `reduce`, follow the order of inputs in the ODS.
  2. Respect varargs for `inputs`, `init_values` and `results` of `reduce`.
  3. Change `x + y` to `add(x, y)`.
  4. Inline `centered_operand_squared` to avoid introducing a new name.
  5. Remove comments in `batch_norm_training` - they don't seem to provide additional context, unlike with `batch_norm_inference`.
  6. Add broadcasts where needed.
  7. Reformat using `pyformat`.

**Review by burmako — CHANGES_REQUESTED** [review]

File: docs/spec_draft.md (hunk 8 above)
Comment (subhankarshah) [reply docs/spec_draft.md:383]: 1. I had seed the order, but I wanted to follow the order in which they appear in MLIR syntax, which is why I used named arguments (will change)
2. I had noticed that as well but I thought single values are treated as array of 1 element
3. I had thought of that as well but I thought people would confuse it with stablehlo add, come to think about it that we want stablehlo add
4. sure
5. sure
6. I thought division by scalar is automatically broadcasted
7. sure

Thanks for the feedback!

Comment (burmako) [reply docs/spec_draft.md:383]: 1a. I think that named arguments are a nice touch because they help understand what each argument means - that's less obvious without looking into the `reduce` spec than e.g. with `add` or `mul`.
1b/2. Let's use the ordering and the types from the spec - I'd say that MLIR implementation details are just implementation details as far as the spec goes.
6. Generally, StableHLO doesn't allow implicit broadcasting. The only two instances that support them are described in #442, and I think I'd like to get rid of them in the future.

File: docs/spec_draft.md [hunk 9]
```diff
@@ -348,6 +349,93 @@ logical operation.
...
+
+Computes mean and variance across all dimensions except for the `feature_index`
+dimension and normalizes the `operand` tensor producing `output`, `batch_mean`
+and `batch_var` tensors. More formally, this operation can be expressed as a
+decomposition to existing StableHLO operations using Python-like syntax as
+follows:
+
+```python
+def compute_mean(operand, feature_index):
+  (sum,) = reduce(
```
Comment (subhankarshah) [inline docs/spec_draft.md:364]: I had first written sum then changed it to aggregate because it is reserved in python, also syntax highlighting makes it blue

File: docs/spec_draft.md [hunk 10]
```diff
@@ -348,6 +349,93 @@ logical operation.
...
+      inputs=[operand],
+      init_values=[0],
+      dimensions=[i for i in range(rank(operand)) if i != feature_index],
+      body=lambda x, y: add(x, y))
+  divisor = constant(num_elements(operand) / dim(operand, feature_index))
+  divisor_bcast = broadcast_in_dim(divisor, [], shape(sum))
+  return divide(sum, divisor_bcast)
+
+def compute_variance(operand, mean, feature_index):
```
Comment (subhankarshah) [inline docs/spec_draft.md:373]: I have taken mean as a parameter in compute_variance, it is ok for this case, but if we want to convey that compute_variance is pure and does not depend on externally computed mean then we can remove it.

File: docs/spec_draft.md [hunk 11]
```diff
@@ -348,6 +349,93 @@ logical operation.
...
+  mean_bcast = broadcast_in_dim(mean, [feature_index], shape(operand))
+  centered_operand = subtract(operand, mean_bcast)
```
Comment (subhankarshah) [inline docs/spec_draft.md:375]: would it be better if we wrote `stablehlo.broadcast_in_dim` and `stablehlo.subtract` etc
this was a suggestion from @sdasgup3

File: docs/spec_draft.md (hunk 8 above)
Comment (subhankarshah) [reply docs/spec_draft.md:383]: thanks

File: docs/spec_draft.md (hunk 10 above)
Comment (burmako) [reply docs/spec_draft.md:373]: Let's keep the old version of variance to emphasize that it can be computed from just `operand` and `feature_index`.

File: docs/spec_draft.md (hunk 11 above)
Comment (burmako) [reply docs/spec_draft.md:375]: I think that in the context of a pseudocode in the StableHLO spec it's fine to not qualify these functions with `stablehlo.`. Unless there are naming conflicts with meta-functions like `shape` or `num_elements`, which we don't have at the moment.

File: docs/spec_draft.md (hunk 9 above)
Comment (burmako) [reply docs/spec_draft.md:364]: Let's keep `sum` for the sake of clarity. Calling variables `sum` is not Pythonic, but we only promised "Python-like syntax" above. Also, when Markdown is rendered, the highlighting is gone, which is nice.

**Review by burmako — CHANGES_REQUESTED** [review]

**Review by burmako — APPROVED** [review]

File: docs/spec_draft.md [hunk 12]
```diff
@@ -348,6 +349,94 @@ logical operation.
...
+### Examples
+
+```mlir
+// %operand: [
+//            [[1.0, 2.0], [3.0, 4.0]],
+//            [[3.0, 4.0], [1.0, 2.0]]
+//           ]
+// %scale: [1.0, 1.0]
+// %offset: [1.0, 1.0]
+%results:3 = "stablehlo.batch_norm_training"(%operand, %scale, %offset) {
```
Comment (burmako) [inline docs/spec_draft.md:426]: Nice catch from Sandeep in #351! This should be called `%output:3`, same in the example in the .td file.

**Review by burmako — CHANGES_REQUESTED** [review]

File: docs/spec_draft.md (hunk 12 above)
Comment (subhankarshah) [reply docs/spec_draft.md:426]: Why though? The individual tensor results are named and not a single result named $output / $outputs
```
let results = (outs
      RankedTensorOf<[HLO_Float]>:$output,
      1DTensorOf<[HLO_Float]>:$batch_mean,
      1DTensorOf<[HLO_Float]>:$batch_var);
```

Comment (subhankarshah) [reply docs/spec_draft.md:426]: Can we unpack the results of an operation in MLIR like -
```%output, %batch_mean, %batch_var = "stablehlo.batct_norm_training"(... .......){}```

Comment (subhankarshah) [reply docs/spec_draft.md:426]: Whatever the final decision is, you may LGTM this pr with the comment, I will make the change and merge.

Comment (burmako) [reply docs/spec_draft.md:426]: You're right - that's my oversight. I like the idea of saying `%output, %batch_mean, %batch_var`, then we can refer to the individual results by name in the comment below.

**Review by burmako — APPROVED** [review]

File: docs/spec_draft.md (hunk 12 above)
Comment (sdasgup3) [reply docs/spec_draft.md:426]: LGTM from my side with the fix for the above.

**Review by sdasgup3 — APPROVED** [review]

File: docs/status.md [hunk 13]
```diff
@@ -52,7 +52,7 @@ one of the following tracking labels.
 | atan2                    | no            | yes*         | yes*           | yes             | no          |
 | batch_norm_grad          | no            | yes*         | yes*           | no              | no          |
 | batch_norm_inference     | yes           | revisit      | yes            | no              | no          |
-| batch_norm_training      | no            | yes*         | yes*           | no              | no          |
+| batch_norm_training      | no            | revisit      | yes            | no              | no          |
 | bitcast_convert          | no            | yes*         | infeasible     | yes             | no          |
```
Comment (subhankarshah) [inline docs/status.md:56]: marked revisit like in `batch_norm_inference`

File: docs/spec_draft.md [hunk 14]
```diff
@@ -414,6 +415,95 @@ Numeric precision is implementation-defined.
...
+
+```mlir
+// %operand: [
+//            [[1.0, 2.0], [3.0, 4.0]],
+//            [[3.0, 4.0], [1.0, 2.0]]
+//           ]
+// %scale: [1.0, 1.0]
+// %offset: [1.0, 1.0]
+%output, %batch_mean, %batch_var =
+"stablehlo.batch_norm_training"(%operand, %scale, %offset) {
```
Comment (subhankarshah) [inline docs/spec_draft.md:493]: updated ODS with the same unpacking format

File: docs/status.md (hunk 13 above)
Comment (subhankarshah) [reply docs/status.md:56]: @burmako

File: docs/spec_draft.md (hunk 14 above)
Comment (subhankarshah) [reply docs/spec_draft.md:493]: @burmako

File: docs/status.md (hunk 13 above)
Comment (subhankarshah) [reply docs/status.md:56]: #369

**Comment by subhankarshah** [comment]
@burmako few minor changes for review.

File: docs/spec_draft.md (hunk 14 above)
Comment (burmako) [reply docs/spec_draft.md:493]: Let's keep variable names and the op on the same line? I don't think we split them across multiple lines anywhere in the spec.

File: docs/status.md (hunk 13 above)
Comment (burmako) [reply docs/status.md:56]: Nice catch! Indeed, this should be `revisit`.

**Review by burmako — CHANGES_REQUESTED** [review]

**Review by burmako — APPROVED** [review]
